### PR TITLE
Populate TagInfo with rich metadata using git tag --format

### LINKED ISSUE #955: Tag::List should populate TagInfo with rich metadata using git tag --format
## Summary

`Git::Commands::Tag::List` currently returns `TagInfo` objects with only the `name` field populated. All other fields (`sha`, `objecttype`, `tagger_name`, `tagger_email`, `tagger_date`, `message`) are set to `nil`.

This limitation affects commands that rely on `Tag::List`, such as `Tag::Delete`, which returns `TagInfo` objects with minimal metadata in `TagDeleteResult.deleted`.

## Motivation

Having rich metadata in `TagInfo` provides:
- **Better UX**: Callers can immediately access tag SHA, type, tagger information, and message
- **Consistency**: Matches the pattern of other `-Info` objects (BranchInfo, StashInfo) which have full metadata
- **Utility**: Enables filtering, sorting, and displaying tag information without additional git calls

## Current Implementation

```ruby
# In lib/git/commands/tag/list.rb
def build_tag_info(name)
  Git::TagInfo.new(
    name: name, sha: nil, objecttype: nil,
    tagger_name: nil, tagger_email: nil, tagger_date: nil, message: nil
  )
end
```

The code even has a note acknowledging this:
> Currently only populates the name field. Other metadata fields are set to nil. A future enhancement could use git tag --format to populate these fields.

## Proposed Solution

Use `git tag --list --format` to fetch rich metadata in a single command:

```bash
git tag --list --format='%(refname:short)|%(objectname)|%(objecttype)|%(taggername)|%(taggeremail)|%(taggerdate:iso8601)|%(contents:subject)'
```

Example output:
```
v1.0.0|abc123def456|tag|John Doe|john@example.com|2024-01-15T10:30:00-08:00|Release version 1.0.0
v2.0.0|def456abc123|commit||||
```

### Implementation Steps

1. **Update `ARGS` definition** in `Tag::List` to optionally include `--format` flag
2. **Add `FORMAT_STRING` constant** with the format specification
3. **Update `parse_tags` method** to parse delimited output instead of plain tag names
4. **Update `build_tag_info` method** to populate all TagInfo fields from parsed data
5. **Handle both lightweight and annotated tags** (annotated tags have tagger info, lightweight don't)
6. **Update unit tests** to verify all fields are populated correctly
7. **Update integration tests** to verify real git behavior with both tag types

### Edge Cases to Consider

- Lightweight tags: `tagger_*` and `message` fields will be empty
- Tag names with special characters (unicode, slashes, etc.)
- Tags with multi-line messages (only use subject line)
- Tags with no message

### Backward Compatibility

✅ This is backward compatible - the return type stays `Array<TagInfo>`, just with more fields populated. Existing code that only uses `.name` will continue to work.

## Related

- Issue #947 - Command classes should return -Info data objects (parent tracking issue)
- `Tag::Delete` implementation already expects and uses `TagInfo` objects from `Tag::List`

## Implementation Guidance

**For AI agents**: Follow the [Development Workflow](.github/copilot-instructions.md#development-workflow) in `.github/copilot-instructions.md`:
- Phase 0: TRIAGE - Review this issue and git documentation
- Phase 1: PREPARE - Design the format string and parsing logic
- Phase 2: EXECUTE - Implement TDD style (write tests first, then implementation)
- Phase 3: FINALIZE - Create PR following conventional commits

Key points:
- Use TDD: Write unit tests first to verify parsing logic
- Add integration tests to verify real git behavior
- Update existing specs for `Tag::Delete` to remove the "NOTE: Tag::List currently returns minimal metadata" comments
- Ensure both lightweight and annotated tags are handled correctly
- Test with special characters (unicode, slashes) in tag names and messages

### PR DESCRIPTION
**Pull request by jcouball**
## Summary

Implements #955 - `Git::Commands::Tag::List` now returns `TagInfo` objects with rich metadata populated from `git tag --format`.

### Implementation
- **Updated `Tag::List` command** to use `git tag --format` with unit separator (`\x1f`) delimited fields for safe parsing
- **Added format string** with 7 fields: name, SHA, object type, tagger name, tagger email, tagger date (ISO 8601 strict), and message
- **Added field count validation** to raise `Git::UnexpectedResultError` with helpful error messages for unexpected output
- **Added parsing methods** (`parse_tag_line`, `build_tag_info`, `parse_optional_field`, `parse_message`) to parse formatted output and populate all `TagInfo` fields
- **Handles both tag types appropriately**:
  - **Annotated tags**: All fields populated (name, sha, objecttype='tag', tagger info, message)
  - **Lightweight tags**: Basic fields only (name, sha, objecttype=target object type), tagger fields and message set to nil

### Testing
- **Updated 26 unit tests** to verify formatted output parsing with unit separator delimiter
- **Added error handling test** for malformed output validation
- **Added comprehensive edge case tests** for unicode, special characters, empty messages
- **Created 15 integration tests** with real git commands to verify behavior

## Format String

The command uses a unit separator (`\x1f`) to safely delimit fields (avoiding conflicts with pipes in tag messages):

```
%(refname:short)<US>%(objectname)<US>%(objecttype)<US>%(taggername)<US>%(taggeremail)<US>%(taggerdate:iso8601-strict)<US>%(contents:subject)
```

Where `<US>` represents the unit separator character (`\x1f`).

## Example Output

**Annotated tag:**
```ruby
tag = TagInfo.new(
  name: 'v1.0.0',
  sha: 'abc123def456',
  objecttype: 'tag',
  tagger_name: 'John Doe',
  tagger_email: '<john@example.com>',
  tagger_date: '2024-01-15T10:30:00-08:00',
  message: 'Release version 1.0.0'
)
```

**Lightweight tag:**
```ruby
tag = TagInfo.new(
  name: 'v2.0.0',
  sha: 'def456abc123',
  objecttype: 'commit',  # or 'tree', 'blob', etc. depending on target
  tagger_name: nil,
  tagger_email: nil,
  tagger_date: nil,
  message: nil
)
```

## Backward Compatibility

✅ **Fully backward compatible** - return type remains `Array<TagInfo>`, just with more fields populated. Existing code using only `.name` continues to work without changes.

## Benefits

- **Better UX**: Callers can immediately access tag SHA, type, tagger information, and message
- **Consistency**: Matches the pattern of other `-Info` objects (BranchInfo, StashInfo) which use unit separator delimiter
- **Safety**: Unit separator delimiter prevents issues with pipes in tag messages or names
- **Robustness**: Field count validation provides clear error messages for unexpected output
- **Efficiency**: Single git command retrieves all metadata (no additional git calls needed)
- **Utility**: Enables filtering, sorting, and displaying tag information without extra overhead

## Test Results

- ✅ All 27 unit tests pass
- ✅ All 15 integration tests pass  
- ✅ All 949 RSpec examples pass
- ✅ RuboCop clean

Closes #955

### CODE REVIEW

File: spec/unit/git/commands/tag/list_spec.rb [hunk 1]
```diff
@@ -112,45 +119,136 @@
...
+        expect(result.first.message).to eq('Release 🎉')
+      end
+
+      it 'handles empty message for annotated tags' do
+        no_message_output = ['v1.0|abc123|tag|Dev|<dev@example.com>|2024-01-01T00:00:00Z|']
+        expect(execution_context).to receive(:command_lines).with('tag', '--list',
+                                                                  format_arg).and_return(no_message_output)
+        result = command.call
+        expect(result.first.message).to be_nil
+      end
```
[the remark this answers is not in the document]
Comment (jcouball) [reply spec/unit/git/commands/tag/list_spec.rb:259]: Will switch to non-printable separator.
